### DISCUSSION TITLE
Use Remark42 on a Gatsby (React) site

### DISCUSSION THREAD

**Discussion by BenRoe** [umputun/remark42#1220] [Q&A] [answered]
Hello,
i try to include Remark42 in [my Gatsby site](https://keycapsss.com/help/).

I tried different examples. One from the #753 issue, [this one](https://nehalist.io/adding-commento-to-react-apps-like-gatsby/) and [this](https://koalatea.io/gatsby-add-comments/).

Always the same problem. **Remark42 is included if i open the site, but if i change the page** (click on a link in the sidebar (not shop))**, Remark42 is not loaded.**

I add Remak42 with this`<Comments commentsId="PAGETITLE" />` the the pages (component code below).

Here the docker vars:
| VAR |  |
| --- | --- |
| REMARK_URL | https://comments.keycapsss.com |
| SITE | keycapsss.com |

```js
// comments.js

import React, { useEffect } from 'react'
import PropTypes from 'prop-types'

// Helper to add scripts to our page
const insertScript = (id, parentElement) => {
  const script = window.document.createElement('script')
  script.type = 'text/javascript'
  script.async = true
  script.id = id
  script.innerHTML = `
    var remark_config = {
    host: 'https://comments.keycapsss.com',
    site_id: 'keycapsss.com',
    //url: 'PAGE_URL', // optional param; if it isn't defined
                     // 'window.location.origin + window.location.pathname' will be used
                     //
                     // Note that if you use query parameters as significant part of URL
                     // (the one that actually changes content on page)
                     // you will have to configure URL manually to keep query params, as
                     // 'window.location.origin + window.location.pathname' doesn't contain query params and
                     // hash. For example, default URL for 'https://example/com/example-post?id=1#hash'
                     // would be 'https://example/com/example-post'
                     //
                     // The problem with query params is that they often contain useless params added by
                     // various trackers (utm params) and doesn't have defined order, so Remark42 treats differently
                     // all this examples:
                     // https://example.com/?postid=1&date=2007-02-11
                     // https://example.com/?date=2007-02-11&postid=1
                     // https://example.com/?date=2007-02-11&postid=1&utm_source=google
                     //
                     // If you deal with query parameters make sure you pass only significant part of it
                     // in well defined order
    max_shown_comments: 10, // optional param; if it isn't defined default value (15) will be used
    theme: 'light', // optional param; if it isn't defined default value ('light') will be used
    // page_title: 'Moving to Remark42', // optional param; if it isn't defined 'document.title' will be used
    locale: 'en', // set up locale and language, if it isn't defined default value ('en') will be used
    show_email_subscription: false, // optional param; by default it is 'true' and you can see email subscription feature
                                    // in interface when enable it from backend side
                                    // if you set this param in 'false' you will get notifications email notifications as admin
                                    // but your users won't have interface for subscription
    //simple_view: true, // optional param; overrides the parameter from the backend
                        // minimized UI with basic info only
  };
  !function(e,n){for(var o=0;o<e.length;o++){var r=n.createElement("script"),c=".js",d=n.head||n.body;"noModule"in r?(r.type="module",c=".mjs"):r.async=!0,r.defer=!0,r.src=remark_config.host+"/web/"+e[o]+c,d.appendChild(r)}}(remark_config.components||["embed"],document);
  `
  parentElement.appendChild(script)
  return script
}

// Helper to remove scripts from our page
const removeScript = (id, parentElement) => {
  const script = window.document.getElementById(id)
  if (script) {
    parentElement.removeChild(script)
  }
}

const Comments = ({ commentsId }) => {
  useEffect(() => {
    // If there's no window there's nothing to do for us
    if (!window) {
      return
    }
    const document = window.document
    // In case our #remark42 container exists we can add our commento script
    if (document.getElementById('remark42')) {
      insertScript('comments-script', document.body)
    }

    // Cleanup when the component unmounts
    return () => removeScript('comments-script', document.body)
  }, [commentsId])

  return (
    <>
      <h2>Comments</h2>
      <div id="remark42"></div>
    </>
  )
}

Comments.propTypes = {
  commentsId: PropTypes.string,
}

export default Comments
```

**Comment by umputun** [comment, marked as answer]
_moved to discussions_

@akellbl4 & @paskal  - I'm not sure what the issues here are, but the backed seems to be configured correctly to me. Can you pls take a look? I also think we may want to add the "integration with Gatsby" chapter to our docs as this thing is popular enough to have a dedicated page.

**Reply by olegedly** [reply]
Yes please! I'm currently trying to do it and struggling

EDIT:

First of all, thanks to @BenRoe for providing his code, it helped me with at least getting the comments to work at all.

Secondly, I can now confirm this behavior: on first page load the comments show up, but on changing the page they don't reappear and stay absent. This probably has to do with how Gatsby switches pages without actually reloading the window. If I manually refresh the page (F5) the comments do reappear.

The question is: how can we trigger this proper refresh when we navigate through pages the Gatsby way?

**Reply by olegedly** [reply]
https://stackoverflow.com/questions/61274365/allow-component-to-detect-route-change-in-gatsby

The answer might be here, I'm tired will attack this with renewed vigor tomorrow

**Reply by olegedly** [reply]
I think I have it solved. First, add another useEffect:

```js
React.useEffect(() => {
    if (!window) {
      return;
    }
    const remark42 = window.REMARK42;
    if (remark42) {
      remark42.destroy();
      remark42.createInstance(window.remark_config);
    }
  }, [commentsId]);
```
  This destroys and recreates the remark42 instance, leading to the comments widget being properly loaded after Gatsby path change.
  
  And in the config make sure to specify `url` and remove the trailing slash if present:
  
```js
const insertScript = (id, parentElement) => {
  const script = window.document.createElement("script");
  script.type = "text/javascript";
  script.async = true;
  script.id = id;
  let url = window.location.origin + window.location.pathname;
  if(url.endsWith("/")) {
    url = url.slice(0, -1);
  }
  script.innerHTML = `
    var remark_config = {
      host: "my_url",
      site_id: "my_id",
      url: "${url}",
      theme: "dark",
      components: ["embed"],
    };
    !function(e,n){for(var o=0;o<e.length;o++){var r=n.createElement("script"),c=".js",d=n.head||n.body;"noModule"in r?(r.type="module",c=".mjs"):r.async=!0,r.defer=!0,r.src=remark_config.host+"/web/"+e[o]+c,d.appendChild(r)}}(remark_config.components||["embed"],document);`;
  parentElement.appendChild(script);
  return script;
};
```

Because otherwise the comments for paths with/without the trailing slash are stored separately in the BoltDB. When following a Gatsby Link a page is loaded without the trailing slash, but when refreshing the page (F5) it is loaded with the trailing slash. So essentially every URL can become duplicated in the DB and you may not see your comments from the inverse URL at your present URL. Making sure url is provided without the trailing slash in the remark42 config solves this.

Code can be further optimized, but at least this works.

**Reply by paskal** [reply]
@elmaester, if you would be able to write down steps you need to take to make remark42 work with Gatsby, it would be _extremely_ helpful. I think creating a new folder with `index.md` file in it [here](https://github.com/umputun/remark42/tree/master/site/src/docs/getting-started) would be an appropriate place for such new documentation.

**Reply by akellbl4** [reply]
Probably we will need to create examples on out repo and put it Gatsby integration as a working example.

**Reply by olegedly** [reply]
> @elmaester, if you would be able to write down steps you need to take to make remark42 work with Gatsby, it would be _extremely_ helpful. I think creating a new folder with `index.md` file in it [here](https://github.com/umputun/remark42/tree/master/site/src/docs/getting-started) would be an appropriate place for such new documentation.

Ok, my pleasure. I'll do it soon. I'm using Remark42, so I'm invested in its success now ☺️

**Reply by olegedly** [reply]
I have added a pull request: https://github.com/umputun/remark42/pull/1222

**Reply by paskal** [reply]
Published as https://remark42.com/docs/manuals/integration-with-gatsby/
